### PR TITLE
feat(milestones): cancel/un-cancel milestones with a Cancelled status (DEV-523)

### PR DESCRIPTION
**Pull request by brunodmsi**
## Overview

Frontend for DEV-523. Admins/staff can **Cancel** a pending or past-due milestone from the Milestones Review page (keeps it, badged Cancelled) instead of deleting it, and **Un-cancel** to restore it.

## Solution

- **FE-signed**, matching the existing verify flow: cancel attests a `MilestoneCompleted` with `type: 'cancelled'` via `GapContract.multiAttest` and notifies the indexer; un-cancel revokes the indexer-surfaced cancellation attestation.
- `useMilestoneCancellation` hook — React Query `useMutation` with optimistic updates; blocks cancelling a completed/verified milestone.
- New **Cancelled** review status; `getMilestoneStatus` treats it as terminal (wins over completion/verification). Cancel (confirm dialog + optional reason) and Un-cancel live in the milestone overflow menu; Complete/Verify are hidden and a cancellation banner (who/why) is shown when cancelled.
- Cancelled milestones drop out of the progress bar and the past-due/pending filter; empty filter tabs are now hidden.

## Testing

- `tsc --noEmit` clean, Biome clean, **Next build passes**.
- Unit tests: `getMilestoneStatus` cancelled/terminal cases; hook mutation tests (attests cancellation + notifies, blocks completed, revokes the surfaced uid for un-cancel, errors when no uid).

## Dependencies

Requires **@show-karma/karma-gap-sdk@0.4.28** (karma-gap-sdk#137) for the `cancelled` type — the lockfile bump lands once 0.4.28 is published. Pairs with the gap-indexer PR (indexing + terminality + count exclusion). Joint E2E on a shared preview before merge.

## Requested due date

2026-07-15 (Filecoin / Erin O'Connor).

## Summary by CodeRabbit

* **New Features**
  * Added support for cancelling and uncancelling milestones, including optional cancellation reasons.
  * Added cancellation status badges, tooltips, filters, banners, and administrative actions.
* **Bug Fixes**
  * Cancelled milestones are excluded from pending lists, progress calculations, completion percentages, and milestone totals.
  * Cancelled milestones remain terminal and are no longer marked past due.
* **Tests**
  * Added comprehensive coverage for cancellation, uncancellation, status handling, filtering, and display behavior.

### CODE REVIEW

**Comment by brunodmsi** [comment]
### SDK dependency: staying on published `0.4.27`

Reverted `package.json` from `0.4.28` back to the published `0.4.27` (commit `ead008909`). The bump was the only thing making CI red (`ERR_PNPM_OUTDATED_LOCKFILE` — manifest `0.4.28` vs lockfile `0.4.27`), and it turns out this PR doesn't need it:

- The cancel attestation is built via `sanitizeObject({ type: "cancelled", ... })`, which is typed `any` — the SDK's `type` union never constrains it.
- We call `GapContract.multiAttest` / `multiRevoke` directly and **never** use the SDK's new `Milestone.cancel()` entity method.
- `MilestoneReviewStatus.Cancelled` is a **front-end-local** enum, not the SDK type.

Verified locally against the *real* published `0.4.27` (unpatched union): `tsc --noEmit` clean and `next build` → **✓ Compiled successfully**.

Net effect: this PR is decoupled from the SDK release, so all three DEV-523 PRs are independently mergeable. The SDK PR (karma-gap-sdk#137) still publishes `0.4.28` with the proper `cancelled` typings on its own timeline; a later FE bump can adopt it for stricter compile-time checking.

File: components/GrantCard.tsx [hunk 1]
```diff
@@ -150,8 +155,9 @@ const GrantCardContent = ({
               <div className="flex w-full flex-row flex-wrap justify-start gap-1">
                 <div className="flex h-max w-max items-center justify-start rounded-full bg-slate-50   dark:bg-slate-700 text-slate-600 dark:text-gray-300 px-3 py-1 max-2xl:px-2">
                   <p className="text-center text-sm font-semibold text-slate-600 dark:text-slate-100 max-2xl:text-[13px]">
-                    {completedMilestonesCount(grant.milestones)}/{grant.milestones?.length || 0}{" "}
-                    {pluralize("Milestone", grant.milestones?.length || 0)}
+                    {completedMilestonesCount(grant.milestones)}/
+                    {activeMilestones(grant.milestones).length}{" "}
+                    {pluralize("Milestone", activeMilestones(grant.milestones).length)}
                   </p>
                 </div>
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply components/GrantCard.tsx]: Done in 5f28e698a — the milestone count pill now renders only when `activeMilestones(grant.milestones).length > 0`, so a grant with no active milestones no longer shows "0/0 Milestones". Updated the GrantCard edge-case test accordingly.

File: components/Pages/Admin/MilestonesReview/index.tsx [hunk 2]
```diff
@@ -1006,6 +1042,11 @@ function MilestonesReviewPageContent({
                       onRequestChanges={handleRequestChanges}
                       onDeleteMilestone={handleDeleteMilestone}
                       isDeleting={isDeleting && deletingMilestoneId === selectedMilestone.uid}
+                      canCancelMilestones={canEditOrDeleteMilestones}
+                      onCancelMilestone={handleCancelMilestone}
+                      onUncancelMilestone={handleUncancelMilestone}
+                      isCancelling={isCancelling}
+                      isUncancelling={isUncancelling}
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply components/Pages/Admin/MilestonesReview/index.tsx:1030]: Done in 5f28e698a — added `isStaff` to `isAuthorized`. Staff already had `canCancelMilestones` (via `canEditOrDeleteMilestones = isStaff || hasAdminAccess`) but were blocked at the page gate before reaching the controls.

File: components/Pages/Admin/MilestonesReview/MilestoneCard.tsx [hunk 3]
```diff
@@ -307,6 +331,21 @@ export function MilestoneCard({
...
+
+  const handleConfirmCancel = useCallback(async () => {
+    await onCancelMilestone?.(milestone, cancelReason.trim() || undefined);
+    setCancelReason("");
+  }, [onCancelMilestone, milestone, cancelReason]);
+
+  const handleUncancelMenuClick = useCallback(() => {
+    setIsOverflowOpen(false);
+    onUncancelMilestone?.(milestone);
+  }, [onUncancelMilestone, milestone]);
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply components/Pages/Admin/MilestonesReview/MilestoneCard.tsx]: Done in 5f28e698a — `handleUncancelMenuClick` is now `async` and `await`s `onUncancelMilestone` inside try/catch. The un-cancel mutation's `onError` still owns the user-facing error state; the catch only prevents an unhandled rejection.

File: components/Pages/Admin/MilestonesReview/MilestoneCard.tsx [hunk 4]
```diff
@@ -425,54 +464,129 @@ export function MilestoneCard({
...
-                    role="menu"
+          {(() => {
+            const showDeleteAction = canDeleteMilestones && !hasCompletion;
+            const showCancelAction =
+              canCancelMilestones &&
+              !!onCancelMilestone &&
+              !isCancelled &&
+              !hasCompletion &&
+              !isVerified;
+            const showUncancelAction = canCancelMilestones && !!onUncancelMilestone && isCancelled;
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply components/Pages/Admin/MilestonesReview/MilestoneCard.tsx]: Done in 5f28e698a — `showUncancelAction` now also requires `milestone.cancellation?.uid`. During the optimistic cancel window `uid` is still `""`, so restore is no longer offered until a revocable UID exists.

File: components/Pages/Admin/MilestonesReview/MilestoneCard.tsx [hunk 5]
```diff
@@ -425,54 +464,129 @@ export function MilestoneCard({
...
-                  aria-haspopup="menu"
-                  aria-expanded={isOverflowOpen}
-                >
-                  <EllipsisVerticalIcon className="w-5 h-5 text-gray-500 dark:text-gray-400" />
-                </button>
-                {isOverflowOpen && (
-                  <div
-                    role="menu"
+          {(() => {
+            const showDeleteAction = canDeleteMilestones && !hasCompletion;
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply components/Pages/Admin/MilestonesReview/MilestoneCard.tsx]: Done in 5f28e698a — `showDeleteAction` now includes `&& !isCancelled`, so cancelled milestones can be restored but not permanently deleted, matching the "retained for records" promise.

File: hooks/useMilestoneCancellation.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  const { mixpanel } = useMixpanel();
+
+  const queryKey = QUERY_KEYS.MILESTONES.PROJECT_GRANT_MILESTONES(projectId, programId);
+
+  const notifyAndInvalidate = async (txHash: string | null | undefined, chainId: number) => {
+    if (txHash) {
+      await fetchData(INDEXER.ATTESTATION_LISTENER(txHash, chainId), "POST", {});
+    }
+    await queryClient.invalidateQueries({ queryKey });
+    await queryClient.invalidateQueries({ queryKey: ["reportMilestones"] });
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply hooks/useMilestoneCancellation.ts:70]: Reasoned deferral: this mirrors the established pattern in **every** milestone mutation hook — verify (`useMilestoneCompletionVerification.ts:202`), `useMilestoneEdit`, `useGrantCompletionRevoke`, `useGrant`, and the milestone forms all `await fetchData(ATTESTATION_LISTENER…)` then invalidate without threading the listener error. Making notify/invalidate listener-acknowledgement-aware is a worthwhile improvement, but doing it for cancel alone would diverge it from every sibling flow. Tracking as a repo-wide follow-up rather than a cancel-only change.

File: hooks/useMilestoneCancellation.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  const resolveMilestoneRecipient = async (
+    milestone: GrantMilestoneWithCompletion,
+    data: ProjectGrantMilestonesResponse
+  ): Promise<{ recipient: Hex; gapClient: GAP; walletSigner: Signer }> => {
+    const setup = await setupChainAndWallet({
+      targetChainId: +milestone.chainId,
+      currentChainId: chain?.id,
+      switchChainAsync,
+    });
+    if (!setup) throw new Error("Wallet setup was cancelled");
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply hooks/useMilestoneCancellation.ts]: Done in 5f28e698a — chain-setup aborts now throw a `ChainSetupAbortedError` sentinel. Both `onError` handlers still roll back the optimistic update but `return` early for that sentinel, so the failure isn't re-toasted/re-reported (`setupChainAndWallet` already surfaces it centrally).

File: hooks/useMilestoneCancellation.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    };
+  };
+
+  const cancelMutation = useMutation({
+    mutationFn: async ({ milestone, data, reason }: CancelArgs) => {
+      if (milestone.completionDetails || milestone.verificationDetails) {
+        throw new Error(
+          "This milestone has already been completed or verified and cannot be cancelled."
+        );
+      }
```
[the remark this answers is not in the document]
Comment (brunodmsi) [reply hooks/useMilestoneCancellation.ts:118]: Done in 5f28e698a — the cancel guard now also rejects milestones completed via funding application (`fundingApplicationCompletion`) and already-cancelled milestones (`status === 'cancelled' || cancellation != null`), preventing a second cancellation attestation. Added unit tests for both paths.
